### PR TITLE
[`Add`]: add `IReadOnlyStoreView`

### PR DESCRIPTION
**Pull request by nan01ab**
# Description

Currently `DataCache` is always used here to access the `Store`, and `DataCache` is also used to collect some `Put` and `Delete` operations, then commit these operations to `Store`.

But sometimes, only read-only access to the `Store`, or caching is not need(because  keys just access once), or a global cache is needed(for example: CurrentIndex), so current `DataCache` implementation  may have additional overhead, or some optimization cannot be used.

So, this PR adds `IReadOnlyStoreView` and it's default implementation `ReadOnlyStoreView` for future usages.
And this PR doesn't change any of existing  implementations or usage(for more review-able).

## Type of change

- [x] Optimization (the change is only an optimization)
- [ ] Style (the change is only a code style for better maintenance or standard purpose)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by Jim8y** [comment]
if you check the snapshotcache, you will find a weird thing, which is we use nullable to represents `readonly`.

**Comment by Hecate2** [comment]
> if you check the snapshotcache, you will find a weird thing, which is we use nullable to represents `readonly`.

Cannot understand this sentence

**Comment by Hecate2** [comment]
(So we use `ClonedCache` for writing in a cache (without persistence)?)

File: src/Neo/Persistence/IReadOnlyStoreView.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+// repository or http://www.opensource.org/licenses/mit-license.php
+// for more details.
+//
+// Redistribution and use in source and binary forms with or without
+// modifications are permitted.
+
+using Neo.SmartContract;
+using System.Collections.Generic;
+
+
```
Comment (shargon) [inline src/Neo/Persistence/IReadOnlyStoreView.cs:15]:
```suggestion
```

File: src/Neo/Persistence/ReadOnlyStoreView.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+//
+// Redistribution and use in source and binary forms with or without
+// modifications are permitted.
+
+using Neo.SmartContract;
+using System.Collections.Generic;
+
+namespace Neo.Persistence
+{
+    public class ReadOnlyStoreView : IReadOnlyStoreView
```
Comment (shargon) [inline src/Neo/Persistence/ReadOnlyStoreView.cs:17]: Avoid this class?

File: benchmarks/Neo.Benchmarks/Persistence/Bechmarks_ReadOnlyStoreView.cs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+// Bechmarks_ReadOnlyStoreView.cs file belongs to the neo project and is free
+// software distributed under the MIT software license, see the
+// accompanying file LICENSE in the main directory of the
+// repository or http://www.opensource.org/licenses/mit-license.php
+// for more details.
+//
+// Redistribution and use in source and binary forms with or without
+// modifications are permitted.
+
+
```
Comment (shargon) [inline benchmarks/Neo.Benchmarks/Persistence/Bechmarks_ReadOnlyStoreView.cs:12]:
```suggestion
```

**Comment by Jim8y** [comment]
> > if you check the snapshotcache, you will find a weird thing, which is we use nullable to represents `readonly`.
> 
> Cannot understand this sentence

snapshotcache implicitly can be readonly/non-readonly, up to whether it is initialized with an IReadOnlyStore or an ISnapshot.

**Review by Jim8y — APPROVED** [review]

**Comment by Jim8y** [comment]
But i would like to have another name, cause storeview is always readonly.

**Review by cschuchardt88 — APPROVED** [review]

**Comment by cschuchardt88** [comment]
How about `View` for read-only? `Store` for CRUD? `Cache` for memory/disk saving?

**Review by shargon — APPROVED** [review]

**Pull request merged**